src/wallet/wallet.cpp

```diff
@@ -3038,69 +3001,89 @@ bool CWallet::CreateTransactionInternal(
-                        nFeeRet += additionalFeeNeeded;
-                        break; // Done, able to increase fee from change
+                    change_position->nValue -= nFeeRet;
+
+                    // We want to drop the change to fees if:
+                    // 1. The change output would be dust
+                    // 2. The change is within the (almost) exact match window, i.e. it is less than or equal to the cost of the change output (cost_of_change)
+                    if (IsDust(*change_position, coin_selection_params.m_discard_feerate) || change_position->nValue <= coin_selection_params.m_cost_of_change)
+                    {
+                        nChangePosInOut = -1;
+                        nFeeRet += change_position->nValue;
+                        txNew.vout.erase(change_position);
                     }
-                }
 
-                // If subtracting fee from recipients, we now know what fee we
-                // need to subtract, we have no reason to reselect inputs
-                if (nSubtractFeeFromAmount > 0) {
-                    pick_new_inputs = false;
+                    // Check if the value in that change output was actually enough to cover the transaction fee.
+                    // We do this here because we may have removed the change output, which will have changed the fee that we actually need, and the fee
+                    // that we are actually paying
+                    // Note that this case should never happen because SelectCoins accounts for the fee. It should've failed rather than get to this point.
+                    if (change_and_fee < nFeeRet) {
+                        error = _("Cannot reduce change to cover transaction fee");
+                        return false;
+                    }
                 }
+                // Reduce output values for subtractFeeFromAmount
+                else if (nSubtractFeeFromAmount != 0) {
+                    int i = 0;
+                    bool fFirst = true;
+                    CAmount to_reduce = nFeeRet;
+                    if (nChangePosInOut != -1) {
+                        // Although we aren't going to reduce this change output to handle the transaction fees,
+                        // we still want to check if we would drop it to fees
+                        std::vector<CTxOut>::iterator change_position = txNew.vout.begin()+nChangePosInOut;
+                        if (IsDust(*change_position, coin_selection_params.m_discard_feerate) || change_position->nValue <= coin_selection_params.m_cost_of_change)
+                        {
+                            nChangePosInOut = -1;
+                            nFeeRet += change_position->nValue;
+                            // We need to reduce the amount to subtract from the recipients so they don't overpay
+                            // as dropping the change accounts for some fees.
+                            to_reduce = change_position->nValue;
```

Comment:
Yes, there is a bug here. Removed changing `nFeeRet` and fixed the `to_reduce` calculation.

Indeed, this is supposed to be a discount to the recipients, so it was supposed to be `-=`. Because this will cause `to_reduce` and `nFeeRet` to deviate, it is necessary for `to_reduce` to exist. Otherwise `nFeeRet` will misreport the actual transaction fee for this transaction.